apps/helloshard/contract.sol

```diff
@@ -0,0 +1,291 @@
+pragma solidity >=0.4.22 <0.6.0;
+
+contract MpcCoordinator {
+    /* A blockchain-based MPC coordinator.
+     * 1. Keeps track of the MPC "preprocessing buffer"
+     * 2. Accepts client input 
+     *      (makes use of preprocess randoms)
+     * 3. Initiates MPC epochs (MPC computations)
+     *      (can make use of preprocessing values if needed)
+     */
+
+    // Session parameters
+    uint public n;
+    uint public t;
+    address[] public shard_1;
+    address[] public shard_2;
+    mapping (address => uint) public servermap;
+    // mapping (address => uint) public shard_1_map;
+    // mapping (address => uint) public shard_2_map;
+
+    // Who shards?
+    // ===========
+    // A different approach could be that that a list of servers is
+    // passed and then split into shards, as opposed to expect shards.
+    // 
+    // In other words, the task of assigning servers to shards can be:
+    // 
+    // 1. the responsibility of the code that instantiates the contract
+    // 2. the responsibility of the contract code
+    constructor(address[] _shard_1, address[] _shard_2, uint _t) public {
+        // for simplicity, both shards have the same number of servers
+        require(_shard_1.length == _shard_2.length);
+	    n = _shard_1.length;
+	    t  = _t;
+	    require(3*t < n);
+	    shard_1.length = n;
+	    shard_2.length = n;
+	    for (uint i = 0; i < n; i++) {
+	        shard_1[i] = _shard_1[i];
+	        shard_2[i] = _shard_2[i];
+	        servermap[_shard_1[i]] = i+1;   // servermap is off-by-one
```

Comment:
I see why something like this is necessary. On the other hand, it might not be efficient if we have more than just 2 shards of different lengths. What else could we use here, mapping address => PartyStruct?

Reply:
Yes definitely, it needs to be generalized/improved.

Reply:
Would it make sense to use eth addresses as node ids instead of integers (`i`)?